llm text generation with a range of models (#2208)
openvinotoolkit/model_server: demos/python_demos/llm_text_generation/README.md

## Download model

Download the model using `download_model.py` script:

```bash
cd demos/python_demos/llm_text_generation
pip install -r requirements.txt

python download_model.py --help
INFO:nncf:NNCF initialized successfully. Supported frameworks detected: torch, onnx, openvino
usage: download_model.py [-h] --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mpt-7b-chat,qwen-7b-chat,chatglm3-6b,mistal-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}

Script to download LLM model based on https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/254-llm-chatbot

options:
  -h, --help            show this help message and exit
  --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mpt-7b-chat,qwen-7b-chat,chatglm3-6b,mistal-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}
                        Select the LLM model out of supported list

> python download_model.py --model $(SELECTED_MODEL)

Reviewer: ```suggestion
python download_model.py --model ${SELECTED_MODEL}
```

```
The model will appear in `./tiny-llama-1b-chat` directory.

## Quantization - optional

Quantization can be applied on the original model. It can reduce the model size and memory requirements. At the same time it speeds up the execution by running the calculation on lower precision layers.

```bash
python quantize_model.py --help
INFO:nncf:NNCF initialized successfully. Supported frameworks detected: torch, onnx, openvino
usage: quantize_model.py [-h] --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mpt-7b-chat,qwen-7b-chat,chatglm3-6b,mistal-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}

Script to quantize LLM model based on https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/254-llm-chatbot

options:
  -h, --help            show this help message and exit
  --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mpt-7b-chat,qwen-7b-chat,chatglm3-6b,mistal-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}
                        Select the LLM model out of supported list

> python quantize_model.py --model $(SELECTED_MODEL)

Reviewer: ```suggestion
python quantize_model.py --model ${SELECTED_MODEL}
```

## Download model

Download the model using `download_model.py` script:

```bash
cd demos/python_demos/llm_text_generation
pip install -r requirements.txt

python download_model.py --help
INFO:nncf:NNCF initialized successfully. Supported frameworks detected: torch, onnx, openvino

> usage: download_model.py [-h] --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mpt-7b-chat,qwen-7b-chat,chatglm3-6b,mistal-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}

Reviewer 2: ```suggestion
usage: download_model.py [-h] --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mpt-7b-chat,qwen-7b-chat,chatglm3-6b,mistral-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}
```

Script to download LLM model based on https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/254-llm-chatbot

options:
  -h, --help            show this help message and exit

>   --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mpt-7b-chat,qwen-7b-chat,chatglm3-6b,mistal-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}

Reviewer 2: ```suggestion
  --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mpt-7b-chat,qwen-7b-chat,chatglm3-6b,mistral-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}
```

> INFO:nncf:NNCF initialized successfully. Supported frameworks detected: torch, onnx, openvino
> usage: download_model.py [-h] --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mpt-7b-chat,qwen-7b-chat,chatglm3-6b,mistral-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}
> 
> Script to download LLM model based on https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/254-llm-chatbot
> 
> options:
>   -h, --help            show this help message and exit
>   --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mistral-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}
>                         Select the LLM model out of supported list

Reviewer: ```suggestion
INFO:nncf:NNCF initialized successfully. Supported frameworks detected: torch, onnx, openvino
usage: download_model.py [-h] --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mistral-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}

Script to download LLM model based on https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/254-llm-chatbot

optional arguments:
  -h, --help            show this help message and exit
  --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mistral-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}
                        Select the LLM model out of supported list
```

python download_model.py --model ${SELECTED_MODEL}

```
The model will appear in `./tiny-llama-1b-chat` directory.

> ## Quantization - optional

Reviewer 2: ```suggestion
## Weight Compression - optional
```

Quantization can be applied on the original model. It can reduce the model size and memory requirements. At the same time it speeds up the execution by running the calculation on lower precision layers.

```bash

> python quantize_model.py --help

Reviewer 2: Can we change this to be `compress_model.py` or something like that? @dtrawins  Since this is not full quantization, we should use the weight compression terminology.

INFO:nncf:NNCF initialized successfully. Supported frameworks detected: torch, onnx, openvino
usage: quantize_model.py [-h] --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mpt-7b-chat,qwen-7b-chat,chatglm3-6b,mistal-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}

Script to quantize LLM model based on https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/254-llm-chatbot

options:
  -h, --help            show this help message and exit

>   --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mpt-7b-chat,qwen-7b-chat,chatglm3-6b,mistal-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}

Reviewer 2: ```suggestion
  --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mpt-7b-chat,qwen-7b-chat,chatglm3-6b,mistral-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}
```

> INFO:nncf:NNCF initialized successfully. Supported frameworks detected: torch, onnx, openvino
> usage: compress_model.py [-h] --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mpt-7b-chat,qwen-7b-chat,chatglm3-6b,mistal-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}
> 
> Script to compress LLM model based on https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/254-llm-chatbot
> 
> options:
>   -h, --help            show this help message and exit
>   --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mistral-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}
>                         Select the LLM model out of supported list

Reviewer: ```suggestion
INFO:nncf:NNCF initialized successfully. Supported frameworks detected: torch, onnx, openvino
usage: compress_model.py [-h] --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mistral-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}

Script to compress LLM model based on https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/254-llm-chatbot

optional arguments:
  -h, --help            show this help message and exit
  --model {tiny-llama-1b-chat,red-pajama-3b-chat,llama-2-chat-7b,mistral-7b,zephyr-7b-beta,neural-chat-7b-v3-1,notus-7b-v1,youri-7b-chat}
                        Select the LLM model out of supported list
```

To test unary example:
```bash
docker run -it --rm -p 9000:9000 -v ${PWD}/servable_unary:/workspace -v ${PWD}/${SELECTED_MODEL}:/model openvino/model_server:py -e SELECTED_MODEL=${SELECTED_MODEL} --config_path /workspace/config.json --port 9000
```

To test the streaming example just mount different workspace location from `./servable_stream`.
It contains modified `model.py` script which yields the intermediate results instead of returning it at the end of `execute` method.
The `graph.pbtxt` is also modified to include cycle in order to make the Python Calculator run in a loop.  

```bash
docker run -it --rm -p 9000:9000 -v ${PWD}/servable_stream:/workspace -v ${PWD}/${SELECTED_MODEL}:/model openvino/model_server:py -e SELECTED_MODEL=${SELECTED_MODEL} --config_path /workspace/config.json --port 9000
```

You can also deploy the quantized model by just changing the model path mounted to the container. For example:

```bash

> docker run -it --rm -p 9000:9000 -v ${PWD}/servable_stream:/workspace -v ${PWD}/${SELECTED_MODEL}_INT8_compressed_weights:/model openvino/model_server:py -e SELECTED_MODEL=${SELECTED_MODEL} --config_path /workspace/config.json --port 9000

Reviewer: only the last ovms docker container will be tested with following commands. will it work both for client_unary.py and client_stream.py  examples?

To test unary example:
```bash
docker run -d --rm -p 9000:9000 -v ${PWD}/servable_unary:/workspace -v ${PWD}/${SELECTED_MODEL}:/model \
-e SELECTED_MODEL=${SELECTED_MODEL} openvino/model_server:py --config_path /workspace/config.json --port 9000
```

You can also deploy the compressed model by just changing the model path mounted to the container. For example:

```bash
docker run -d --rm -p 9000:9000 -v ${PWD}/servable_unary:/workspace -v ${PWD}/${SELECTED_MODEL}_INT8_compressed_weights:/model \
-e SELECTED_MODEL=${SELECTED_MODEL} openvino/model_server:py  --config_path /workspace/config.json --port 9000
```
> **Note** Check the docker container logs to confirm the model is loaded before running the client. Depending on the model and hardware it might take up to several minutes.

### Running the client with LLM model and unary gRPC call

Install python client dependencies. This is a common step also for the streaming client.
```bash
pip install -r client_requirements.txt
```

Run time unary client `client_unary.py`:
```bash

> python3 client_unary.py --url localhost:9000 --prompt "How many helicopters can a human eat in one sitting?"

Reviewer 2: @dtrawins  can we change this prompt to something more realistic? I suggest changing helicopters to apples or coming up with a new prompt.

```

Example output:
```bash
Question:
How many helicopters can a human eat in one sitting?

Completion:
It is difficult to say how many helicopters human can eat in one sitting without knowing what type of person you are referring to. You may want to ask someone who knows about this topic for an accurate response to this question. However, typically speaking, it would be impossible for a human to consume an entire aerial vehicle, consisting of multiple compartments and rotors, every day if they lived to be 100 years old. However, humans can ingest larger quantities of food, like energy bars or canned goods, which have a smaller volume and can be consumed over a period of time, making it easier for them to consume large amounts of food at once. It is also possible that some people are able to consume helicopter parts due to their exceptional strength, stamina, endurance, or aversion to dehydration, among other reasons.

> Total time 11662 ms

Reviewer: There is no information about total time in stdout
```suggestion
```

```

## Use LLM model with gRPC streaming

### Deploy OpenVINO Model Server with the Python Calculator

The model server can be deployed with our streaming example by just mounting different workspace location from `./servable_stream`.

> It contains modified `model.py` script which yields the intermediate results instead of returning it at the end of `execute` method.

Reviewer 2: ```suggestion
It contains a modified `model.py` script which provides intermediate results instead of returning the full result at the end of the `execute` method.
```

The `graph.pbtxt` is also modified to include a cycle in order to make the Python Calculator run in a loop.  

```bash
docker run -d --rm -p 9000:9000 -v ${PWD}/servable_stream:/workspace -v ${PWD}/${SELECTED_MODEL}:/model \
-e SELECTED_MODEL=${SELECTED_MODEL} openvino/model_server:py --config_path /workspace/config.json --port 9000
```

> Like with the unary example, you can also deploy the compressed model by just changing the model path mounted to the container. For example:

Reviewer 2: ```suggestion
Just like the unary example, you may deploy the compressed model(s) by simply changing the model path mounted to the container. For example, to deploy the 8-bit weight compressed model:
```

```bash
docker run -d --rm -p 9000:9000 -v ${PWD}/servable_stream:/workspace -v ${PWD}/${SELECTED_MODEL}_INT8_compressed_weights:/model \
-e SELECTED_MODEL=${SELECTED_MODEL} openvino/model_server:py --config_path /workspace/config.json --port 9000
```

> > **Note** Check the docker container logs to confirm the model is loaded before running the client. Depending on the model and hardware it might take up to several minutes.

Reviewer 2: ```suggestion
> **NOTE** Check the Docker container logs to confirm that the model is loaded before sending requests from a client. Depending on the model and hardware it might take a few seconds or several minutes.
```

> ## Running the client with the LLM model and gRPC streaming

Reviewer 2: ```suggestion
## Run a client with the LLM and gRPC streaming
```

Run time streaming client `client_stream.py`:
```bash

> python3 client_stream.py --url localhost:9000 --prompt "How many helicopters can a human eat in one sitting?"

Reviewer 2: @dtrawins can we change this prompt to something more realistic? I suggest replacing helicopters with apples or some food that people would eat. Or create a new prompt.
